keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+* Based on logs generated on master node
+* Aggregate by components:
+  * kube-controller-manager
+  * kube-scheduler
+  * kube-apiserver
+  * kubelet
+* Only for log lines generated by klog (e.g. prefix `I0129 03:26:19.857798       1 garbagecollector.go:517]`)
+* Identified by code line that generated log (e.g. `garbagecollector.go:517`)
+
+We assume that exact details of taking measurements can be improved, but overall methodology is solid. With those criteria we concluded that covering 99.9% of logs on master node are generated by 22 log messages. Exact list of those logs is provided in the detail design section.
+
+#### JSON output format
+
+Introduction of new methods to klog library will make identification of different components of the log message much easier.
+With klog v2 we can take further advantage of this fact and add an option to produce structured logs in JSON format.
+
+Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```go
+klog.Infof("Updated pod %s status to ready", pod.name)
+```
+That would result in log
+
+```json
+{
+   "ts": 1580306777.04728,
+   "level": "info",
+   "msg": "Updated pod kubedns status to ready"
+}
+```
+
+And
+
+```go
+klog.InfoS("Pod status updated", "pod"", metadata.Pod(pod), "status"", "ready")
+```
+
+That would result in log
+```json
+{
+   "ts": 1580306777.04728,
+   "level": "info",
+   "msg": "Pod status updated",
+   "pod":{
+      "name": "nginx-1",
+      "namespace": "default"
+   },
+   "status": "ready"
+}
+```
+
+
+### Logging configuration
+
+To allow selection of the logging output format we would like to introduce a new logging configuration shared by all kubernetes components.
+
+`LoggingConfig` structure should be implemented as part of `k8s.io/component-base` options and include a common set of flags for logger initialization.
+
+`--logging-format` flag should allow you to pick between logging logging output formats. Setting this flag will select a particular logger implementation.
+
+Proposed flag `--logging-format` values:
+* `text` for old text-based logging format
+* `json` for new JSON format
+
+### Migration / Graduation Criteria
+
+#### Alpha
+
+Preparation and migration of selected logs to new logging methods:
+* Support for the new InfoS, WarningS, ErrorS and FatalS method will be implemented in klog and the new version of klog will be released.
+* Kubernetes will be upgraded to use the new klog version.
+* Helper methods for producing consistent Kubernetes resource identifiers to be used in logs will be implemented.
+* Most important klog calls will be identified and will be manually migrated from using format based to structured methods.
```

Comment:
This effort should also trigger a revision of https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md

And a community wide communication to gain some developer awareness could have a good impact.  If developers are aware of the new calls and standards then they may make improvements to the logging in various components as part of other maintenance activities.

Reply:
Very good suggestion. We will definitely want to update this doc.

Reply:
Added to beta step